### PR TITLE
External accounts implemented for issue #5640

### PR DESCRIPTION
**Pull request by r001**
External acccounts allow Metamask to track account whose private keys are stored on a safe device, and tx data and signatures are exchanged by using QR codes. This allows very safe account password storage. 

This pull request enables users to create so called external accounts. This means that after applying this pull request in the "import account" menu there will be an "external" tab that enables importing an ethereum address. There is no local private key stored, but upon confirming a transaction a QR code is presented and the transaction can be signed by an external signer. And the signature can be provided in the signature input field. By providing the signature, the transaction will be signed, and published. 

The way it works is the following:
1. User creates transaction (or msg signing) and Confirms it.
2. Request goes to the background, and since the from-accunt belongs to eth-external-account-keyring, the keyrings `signTransaction()` function is called.
3. Sign transaction adds an object to eth-keyring-controller's `extToSign` array, which is monitored by the UI modal.js base component.
4. Since there is a new `extToSign` entry `ui/app/components/modals/modal.js` starts the external-sign-modal.js window, and displays QR code, and requests signature. 
5. Once signature received from user, the tx is segned normall, and handled in a standard way by Metamask.

To allow for external signatures, **eth-keyring-controller** package had to be updated, and **eth-external-account-keyring** package was created. So in the package.json my updated ones are the sources.
Eth-external-account-keyring is available at:
https://github.com/r001/eth-external-account-keyring

Install instructions for reference implementation of encoding/decoding QR code can be found at 
https://github.com/r001/qr-encoding/blob/master/INSTALL.md

Modified signer that can create appropriate signature can be found at :
https://github.com/r001/dapptools/tree/ethsign-signature
To create signature use `ethsign tx --signature-only <tx options>...`

### CODE REVIEW

**Comment by r001** [comment]
@caffeinum Cold you pls reintroduce your issues here about visibility. I have rebased the previous pr on top of 5.3.0, and github has not recognized the delete, push again cycle of mine.

**Comment by caffeinum** [comment]
Of course, here are my objections: r001/qr-encoding#1

The main idea: I think it could be wise to use human-readable parameters for an encoded link instead of raw bytes.

**Comment by r001** [comment]
@danfinlay I had to rebase pr #5903, so I have reopened here. So you can check it whenever you have time.

**Comment by r001** [comment]
@caffeinum If it turns out we have to go with the readable format as you have suggested. Do you have time to adjust metamask code, and decoder code? BTW I still think it helps not much, since user can not change anything, since we only return signature.

**Comment by caffeinum** [comment]
@r001 currently, I am not much available, as there are other projects, but I will look into what I can do when I will have some extra time!

**Review by tmashuang — CHANGES_REQUESTED** [review]

File: ui/app/ducks/confirm-transaction.duck.js [hunk 1]
```diff
@@ -65,6 +66,7 @@ const initState = {
   nonce: '',
   toSmartContract: false,
   fetchingData: false,
+  signature: '',
```
[the remark this answers is not in the document]
Comment (r001) [reply ui/app/ducks/confirm-transaction.duck.js]: I did the requested changes in 2263fc3cc9db24ad1b94e46591c38849ace8bc01

File: ui/app/components/send/send-content/send-row-wrapper/send-row-error-message/send-row-error-message.component.js [hunk 2]
```diff
@@ -6,16 +6,17 @@ export default class SendRowErrorMessage extends Component {
...
   static contextTypes = {
     t: PropTypes.func,
   };
 
   render () {
-    const { errors, errorType } = this.props
+    const { errors, errorType, customErrors } = this.props
 
-    const errorMessage = errors[errorType]
+    const errorMessage = errors[errorType] || customErrors[errorType]
```
[the remark this answers is not in the document]
Comment (r001) [reply ui/app/components/send/send-content/send-row-wrapper/send-row-error-message/send-row-error-message.component.js]: I did the requested changes in 3bf96476b00c7cb8c57773b643ef66824364965b

File: ui/app/components/modals/modal.js [hunk 3]
```diff
@@ -449,6 +479,18 @@ Modal.prototype.render = function () {
 }
 
 Modal.prototype.componentWillReceiveProps = function (nextProps) {
+  if (
+    nextProps.extToSign.length > this.props.extToSign.length &&
+    //this.props.currentView === 'confTx' &&
```
[the remark this answers is not in the document]
Comment (r001) [reply ui/app/components/modals/modal.js]: Your request was implemented at 69b627cee9006aec77408f09a2aac7df5d85bab8

File: ui/app/components/qr-code.js [hunk 4]
```diff
@@ -42,16 +45,22 @@ QrCodeView.prototype.render = function () {
...
+        value: data,
+        size: width || 480,
+      }),
+    ]),
+
     h(ReadOnlyInput, {
       wrapperClass: 'ellip-address-wrapper',
       inputClass: 'qr-ellip-address',
-      value: checksumAddress(data),
+      value: isDataAddress ? checksumAddress(data) : data,
```
[the remark this answers is not in the document]
Comment (r001) [reply ui/app/components/qr-code.js]: Requested chsnges implemented in e86f2ae500e08996d0aee2a028ec554418eb0f5c

File: ui/app/components/send/send-content/send-row-wrapper/send-row-error-message/send-row-error-message.component.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (r001) [reply ui/app/components/send/send-content/send-row-wrapper/send-row-error-message/send-row-error-message.component.js]: Requested changes in eb061cbff01c958d03b8d6f09bb93933b3134b97

File: ui/app/ducks/confirm-transaction.duck.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (r001) [reply ui/app/ducks/confirm-transaction.duck.js]: Tests seem just fine.

File: ui/app/components/qr-code.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (r001) [reply ui/app/components/qr-code.js]: Tests seem to pass.

File: ui/app/components/send/send-content/send-row-wrapper/send-row-error-message/send-row-error-message.component.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (r001) [reply ui/app/components/send/send-content/send-row-wrapper/send-row-error-message/send-row-error-message.component.js]: Tests are passing. And new ones added.

**Comment by r001** [comment]
@tmashuang Thank you for your help! Meanwhile I could track some additional bugs, and I pushed their fixes too. The `test-deps` fail because I had to use my own updates for `keyring-controller` and added `external-account-keyring` package from my own repo. These need to be put into Metamask's own github place first. 
With  `test-e2e-beta-chrome` and `test-unit` the errors seem to be independent from the implemented new feature.

**Comment by tmashuang** [comment]
@r001 Merge/rebase `develop` should fix the `test-e2e` issue, hopefully also the `test-unit`. After playing with the behavior, it seems that I can not access the QR tx details after I click out of the view. The tx is `pending` /awaiting approval in the transaction list. Would be nice to see the QR details again, if the popup closes while viewing the modal. Maybe adding a button to view the QR again in the `transaction-list-item-details` Also, is the just a proof of concept? How feasible would it be to make to compatible with other signers, i.e. parity signer?

**Comment by r001** [comment]
@r001 Merge/rebase develop should fix the test-e2e issue, hopefully also the test-unit. After playing with the behavior, it seems that I can not access the QR tx details after I click out of the view. The tx is pending /awaiting approval in the transaction list. Would be nice to see the QR details again, if the popup closes while viewing the modal. Maybe adding a button to view the QR again in the transaction-list-item-details Also, is the just a proof of concept? How feasible would it be to make to compatible with other signers, i.e. parity signer?
@tmashuang 
> After playing with the behavior, it seems that I can not access the QR tx details after I click out of the view. The tx is pending /awaiting approval in the transaction list. Would be nice to see the QR details again, if the popup closes while viewing the modal.
I'll give it a try. I was testing this with the whole page version of Metamask, not the popup one.
> Maybe adding a button to view the QR again in the transaction-list-item-details 
If you don't mind, I would only do that after an initial version is merged into `develop`.
> Also, is the just a proof of concept?
No, I want this to be added to Metamask. By the time people will realize that any hardware that is physically connected to the online pc is not the safest solution, Metamask will have a safer alternative. And actually I have been using my old phone already with Metamask to do all my transactions on Ethereum with this new feature. Works like a charm. 
>How feasible would it be to make to compatible with other signers, i.e. parity signer?
Absolutely possible. Currently there is  a custom made encoding implemented instead of RLP that Parity uses. We can do that too. The hard part of integrating the external feature is already done. Making it compatible with Parity signer is just about how we present the same data.

All the tests check out now except for the `test-deps`, and that is because of the custom libs I used for updated `keyring-controller` and `external-account-keyring` package.

**Comment by tmashuang** [comment]
I would prefer the have those official external package version bumped, instead of specific commits. 
I see one of the KeyringController, will take a look and request another person to review it as well, https://github.com/MetaMask/KeyringController/pull/29.

Could you PR `ethereumjs-wallet` with you changes as well?

**Comment by r001** [comment]
> Could you PR `ethereumjs-wallet` with you changes as well?

Makes no sense to PR it since it was an old hackish way of mine trying to introduce external accounts. The original wallet will work just as fine with this version. I have updated package.js accordingly.

**Comment by r001** [comment]
This PR has been reopened in PR #6143

**Pull request closed without merging**